Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -0,0 +1,198 @@
+lemma mem_vertices_cons {a b c : V} (p : Path a b)
+    (e : b ⟶ c) {x : V} :
+    x ∈ (p.cons e).vertices ↔ x ∈ p.vertices ∨ x = c := by
+  simp only [vertices_cons]
+  simp_all only [concat_eq_append, mem_append, mem_cons, not_mem_nil, or_false]
+
+lemma verticesSet_nil {a : V} : {v | v ∈ (nil : Path a a).vertices} = {a} := by
+  simp only [vertices_nil, List.mem_singleton, Set.ext_iff, Set.mem_singleton_iff]
+  exact fun x ↦ Set.mem_setOf
+
+@[simp]
+lemma verticesSet_cons {a b c : V} (p : Path a b) :
+  {v | v ∈ p.vertices ∨ v = c} = {v | v ∈ p.vertices} ∪ {c} := by
+  simp only [Set.union_singleton]
+  ext v
+  simp only [mem_verticesSet_iff, Set.mem_insert_iff]
+  rw [or_comm]; simp
+
+/-- The length of vertices list equals path length plus one -/
+@[simp]
+lemma vertices_length {V : Type*} [Quiver V] {a b : V} (p : Path a b) :
+    p.vertices.length = p.length + 1 := by
+  induction p with
+  | nil => simp only [vertices_nil, List.length_cons, List.length_nil, zero_add, length_nil]
+  | cons p' e ih =>
+    simp only [vertices_cons, length_cons, List.length_concat, ih]
+
+lemma length_vertices_pos {a b : V} (p : Path a b) :
+    0 < p.vertices.length := by simp
+
+lemma vertices_ne_nil {a : V} {b : V} (p : Path a b) : p.vertices ≠ [] := by
+  simp [← List.length_pos_iff_ne_nil]
+
+/-- The head of the vertices list is the start vertex -/
+@[simp]
+lemma vertices_head? {a b : V} (p : Path a b) : p.vertices.head? = some a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head?_cons]
+  | cons p' e ih => simp [ih]
+
+@[simp]
+lemma length_toPath {a b : V} (e : a ⟶ b) : e.toPath.length = 1 := rfl
+
+@[simp]
+lemma vertices_getLast {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_ne_nil) :
+    p.vertices.getLast h = b := by
+  induction p with
+  | nil => simp only [vertices_nil, List.getLast_singleton]
+  | cons p' e ih => simp
+
+@[simp]
+lemma vertices_comp {a b c : V} (p : Path a b) (q : Path b c) :
+  (p.comp q).vertices = p.vertices.dropLast ++ q.vertices := by
+  induction q with
+  | nil => simpa using (List.dropLast_append_getLast p.vertices_ne_nil).symm
+  | cons q' e ih => simp [ih]
+
+lemma start_mem_vertices {a b : V} (p : Path a b) : a ∈ p.vertices := by
+  induction p with
+  | nil => simp
+  | cons p' e ih => simp [ih]
+
+@[simp] lemma length_eq_zero_iff {a : V} (p : Path a a) :
+    p.length = 0 ↔ p = Path.nil := by
+  cases p <;> tauto
+
+/-- The head of the vertices list is the start vertex. -/
+@[simp]
+lemma vertices_head_eq {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_ne_nil) :
+    p.vertices.head h = a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head_cons]
+  | cons p' _ ih => simp [List.head_append_of_ne_nil (vertices_ne_nil p'), ih]
+
+/-- The last element of the vertices list is the end vertex. -/
+lemma vertices_getLast_eq_end {a b : V} (p : Path a b) :
+  p.vertices.getLast (vertices_ne_nil p) = b := by
```

Comment:
Let's drop the `_end` since we directly reference `b` and also add an autoParam for the not-nil condition to make this lemma slightly more useful when rewriting:
```suggestion
lemma vertices_getLast_eq {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_ne_nil) :
  p.vertices.getLast h = b := by
```